### PR TITLE
Feature/empty villages

### PR DESCRIPTION
**Pull request by Raycoms**
Closes #
Closes #
Closes #

# Changes proposed in this pull request:
- Add worldgen feature to spawn empty colonies:
-- Caledonian, Asian, Medieval Oak, and Medieval Dark oak for the time being.
- Colonies can be claimed through the townhall and huts "claimed".
-

Review please

### CODE REVIEW

File: .teamcity/pom.xml [hunk 1]
```diff
@@ -0,0 +1,104 @@
+<?xml version="1.0"?>
```
Comment (someaddons) [inline .teamcity/pom.xml]: shouldnt this stuff be in its own PR?^^

File: src/api/java/com/minecolonies/api/colony/IColonyManager.java [hunk 2]
```diff
@@ -432,4 +432,10 @@ static IColonyManager getInstance()
      * Reset all colony views on login to new world.
      */
     void resetColonyViews();
+
+    /**
+     * Open the new reactivation window.
+     * @param pos the pos to open it at.
+     */
+    void openReactivationWindow(final BlockPos pos);
```
Comment (someaddons) [inline src/api/java/com/minecolonies/api/colony/IColonyManager.java:440]: why is this here in the colony manager?

File: src/api/java/com/minecolonies/api/tileentities/TileEntityColonyBuilding.java [hunk 3]
```diff
@@ -514,4 +522,52 @@ public AbstractContainerMenu createMenu(final int id, @NotNull final Inventory i
...
+     * Reactivate the hut of this tileEntity.
+     * Load the schematic data and set the style correctly.
+     */
+    public void reactivate()
+    {
+        final List<String> tags = new ArrayList<>(this.getPositionedTags().get(BlockPos.ZERO));
+        tags.remove("deactivated");
+        if (!tags.isEmpty())
+        {
+            this.setStyle(tags.get(0));
```
Comment (someaddons) [inline src/api/java/com/minecolonies/api/tileentities/TileEntityColonyBuilding.java:537]: order of tags is not guaranteed, so may want to actually check the right one out instead

File: src/api/java/com/minecolonies/api/tileentities/TileEntityColonyBuilding.java [hunk 4]
```diff
@@ -514,4 +522,52 @@ public AbstractContainerMenu createMenu(final int id, @NotNull final Inventory i
     {
         return new ContainerBuildingInventory(id, inv, colonyId, getBlockPos());
     }
+
+    /**
+     * Reactivate the hut of this tileEntity.
+     * Load the schematic data and set the style correctly.
+     */
+    public void reactivate()
```
Comment (someaddons) [inline src/api/java/com/minecolonies/api/tileentities/TileEntityColonyBuilding.java:531]: So far we've had these kinds of utilities always in the windows/messages, shouldnt we keep that to stay consistent?

File: src/main/java/com/minecolonies/coremod/MineColonies.java [hunk 5]
```diff
@@ -96,6 +122,122 @@ public MineColonies()
 
         InteractionValidatorInitializer.init();
         proxy.setupApi();
+
+        IEventBus modEventBus = FMLJavaModLoadingContext.get().getModEventBus();
```
Comment (someaddons) [inline src/main/java/com/minecolonies/coremod/MineColonies.java:104]: I'd like to move this into its own dedicated worldgen event handler so its in line with most of our existing handlers

File: src/main/java/com/minecolonies/coremod/client/gui/townhall/WindowTownHallColonyManage.java [hunk 6]
```diff
@@ -149,7 +151,13 @@ public void onCreate()
         new VanillaParticleMessage(pos.getX(), pos.getY(), pos.getZ(), ParticleTypes.DRAGON_BREATH).onExecute(null, false);
         Minecraft.getInstance().level.playSound(Minecraft.getInstance().player, new BlockPos(Minecraft.getInstance().player.position()),
           SoundEvents.CAMPFIRE_CRACKLE, SoundSource.AMBIENT, 2.5f, 0.8f);
-        Network.getNetwork().sendToServer(new CreateColonyMessage(pos));
+        boolean claim  = false;
+        final BlockEntity entity = Minecraft.getInstance().level.getBlockEntity(pos);
+        if (entity instanceof final TileEntityColonyBuilding hut)
+        {
+            claim = hut.getPositionedTags().containsKey(BlockPos.ZERO) && hut.getPositionedTags().get(BlockPos.ZERO).contains("deactivated");
```
Comment (someaddons) [inline src/main/java/com/minecolonies/coremod/client/gui/townhall/WindowTownHallColonyManage.java]: constant for deactivated

File: src/main/java/com/minecolonies/coremod/client/gui/townhall/WindowTownHallColonyManage.java [hunk 7]
```diff
@@ -149,7 +151,13 @@ public void onCreate()
         new VanillaParticleMessage(pos.getX(), pos.getY(), pos.getZ(), ParticleTypes.DRAGON_BREATH).onExecute(null, false);
         Minecraft.getInstance().level.playSound(Minecraft.getInstance().player, new BlockPos(Minecraft.getInstance().player.position()),
           SoundEvents.CAMPFIRE_CRACKLE, SoundSource.AMBIENT, 2.5f, 0.8f);
-        Network.getNetwork().sendToServer(new CreateColonyMessage(pos));
+        boolean claim  = false;
```
Comment (someaddons) [inline src/main/java/com/minecolonies/coremod/client/gui/townhall/WindowTownHallColonyManage.java]: different name than claim here, it sounds like chunk claiming but its about restoring a building, maybe reactivated?

File: src/main/java/com/minecolonies/coremod/network/messages/client/CreateColonyMessage.java [hunk 8]
```diff
@@ -35,26 +38,31 @@
      */
     BlockPos townHall;
 
+    boolean claim;
```
Comment (someaddons) [inline src/main/java/com/minecolonies/coremod/network/messages/client/CreateColonyMessage.java:42]: oh I think the best name may be "reactivate", claim is too confusing xD

File: src/main/java/com/minecolonies/coremod/structures/EmptyColonyStructure.java [hunk 9]
```diff
@@ -0,0 +1,109 @@
...
+import net.minecraft.world.level.levelgen.structure.PoolElementStructurePiece;
+import net.minecraft.world.level.levelgen.structure.PostPlacementProcessor;
+import net.minecraft.world.level.levelgen.structure.pieces.PieceGenerator;
+import net.minecraft.world.level.levelgen.structure.pieces.PieceGeneratorSupplier;
+import net.minecraftforge.event.world.StructureSpawnListGatherEvent;
+
+import java.util.List;
+import java.util.Optional;
+
+public class EmptyColonyStructure extends StructureFeature<JigsawConfiguration>
```
Comment (someaddons) [inline src/main/java/com/minecolonies/coremod/structures/EmptyColonyStructure.java:32]: javadoc in this class is missing

**Review by someaddons — COMMENTED** [review]
Left some minor comments, one thing that we're probably missing is a config to set spacing and frequency, usually structure spawn mods always have those so you can adjust it

File: src/api/java/com/minecolonies/api/colony/IColonyManager.java (hunk 2 above)
Comment (Raycoms) [reply src/api/java/com/minecolonies/api/colony/IColonyManager.java:440]: Because of api/main seperation, that is the most accessible place from AbstractBlockHut I could find.

File: src/api/java/com/minecolonies/api/tileentities/TileEntityColonyBuilding.java (hunk 3 above)
Comment (Raycoms) [reply src/api/java/com/minecolonies/api/tileentities/TileEntityColonyBuilding.java:537]: I rely on the hut blocks having only 2 tags here. the deactivated and the other, there is no other way to detect it else

File: src/api/java/com/minecolonies/api/tileentities/TileEntityColonyBuilding.java (hunk 4 above)
Comment (Raycoms) [reply src/api/java/com/minecolonies/api/tileentities/TileEntityColonyBuilding.java:531]: I put it here so that it has full access to the variables of the tileEntity

File: src/api/java/com/minecolonies/api/tileentities/TileEntityColonyBuilding.java (hunk 3 above)
Comment (someaddons) [reply src/api/java/com/minecolonies/api/tileentities/TileEntityColonyBuilding.java:537]: could've given it a style prefix

File: src/api/java/com/minecolonies/api/tileentities/TileEntityColonyBuilding.java (hunk 4 above)
Comment (Raycoms) [reply src/api/java/com/minecolonies/api/tileentities/TileEntityColonyBuilding.java:531]: Also, because it's called from 2 different messages

**Review by someaddons — APPROVED** [review]
